src/Compiler/Microsoft.AspNetCore.Razor.Language/src/TagHelperCollector.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Razor.Language;
+
+public abstract partial class TagHelperCollector<T>
+    where T : ITagHelperDescriptorProvider
+{
+    // This type is generic to ensure that each descendent gets its own instance of this field.
+    private static readonly ConditionalWeakTable<IAssemblySymbol, Cache> s_perAssemblyCaches = new();
+
+    private readonly Compilation _compilation;
+    private readonly ISymbol? _targetSymbol;
+
+    protected TagHelperCollector(Compilation compilation, ISymbol? targetSymbol)
+    {
+        _compilation = compilation;
+        _targetSymbol = targetSymbol;
+    }
+
+    private static bool IsTagHelperAssembly(IAssemblySymbol assembly)
+    {
+        // This as a simple yet high-value optimization that excludes the vast majority of
+        // assemblies that (by definition) can't contain a component.
+        return assembly.Name != null && !assembly.Name.StartsWith("System.", StringComparison.Ordinal);
```

Comment:
Suggested this offline as well, but it seems like we can make this way more precise: if the assembly is `Microsoft.AspNetCore.Razor` or doesn't have a reference to that assembly, then it cannot possibly have tag helpers, since that's where the `TagHelper` type is defined. This should significantly limit the possible assemblies another order of magnitude.

Reply:
Actually, this won't work because each tag helper provider has more specific requirements than just `ITagHelper`. There's also components, event handlers, etc., which don't require `ITagHelper`. Since this particular filter needs to be a more permissive, I'll just switch this old check around to look for assemblies that start with "Microsoft.AspNetCore." or reference assemblies that start with "Microsoft.AspNetCore.". That should do a better job for us.